File: sample/unix_prog/sysinfo/passwd.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#ifdef __linux__
#include <crypt.h>
#endif

void output(const char *prefix, char *encrypted, int saltlen) {
	if(encrypted == NULL)
		printf("%7s: Not available\n", prefix);
	printf("%7s: %2zd %s\n", prefix, strlen(encrypted) - saltlen, encrypted);
	return;
}

char *
salt(int n, char *salt) {
	int i;
	static char ch[] =
		"abcdefghijklmnopqrstuvwxyz"
		"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
		"1234567890./";
	if(salt == NULL)
		return salt;
	for(i = 0; i < n; i++) {
		salt[i] = ch[rand() % sizeof(ch)];
	}
	return salt;
}

int
main(int argc, char *argv[]) {
	char *ptr, p[64];
	char *p_des, s_des[64] = "";
	char *p_md5, s_md5[64] = "$1$";
	char *p_sha256, s_sha256[64] = "$5$";
	char *p_sha512, s_sha512[64] = "$6$";

	srand(time(0) ^ getpid());

	printf("Password Demo\n");
	if((ptr = getpass("Enter your password: ")) == NULL) {
		perror("getpass");
		return -1;
	}

	strncpy(p, ptr, sizeof(p));
	printf("Your password is: [%s]\n", p);

	if(argc > 1) {
		strncpy(s_des, argv[1], 2);
		strncat(s_md5, argv[1], 8);
		strncat(s_sha256, argv[1], 8);
		strncat(s_sha512, argv[1], 8);
	} else {
		char s[16];
		strncpy(s_des, salt(2, s), 2);
		strncat(s_md5, salt(8, s), 8);
		strncat(s_sha256, salt(8, s), 8);
		strncat(s_sha512, salt(8, s), 8);
	}
	strcat(s_md5, "$");
	strcat(s_sha256, "$");
	strcat(s_sha512, "$");

	p_des = strdup(crypt(p, s_des) ? : "<DES failed>");
	p_md5 = strdup(crypt(p, s_md5) ? : "<MD5 failed>");
	p_sha256 = strdup(crypt(p, s_sha256) ? : "<SHA256 failed>");
	p_sha512 = strdup(crypt(p, s_sha512) ? : "<SHA512 failed>");

	output("DES", p_des, 2);
	output("MD5", p_md5, 12);
	output("SHA256", p_sha256, 12);
	output("SHA512", p_sha512, 12);

	return 0;
}

File: sample/unix_prog/sysinfo/passwd.py
#!/usr/bin/env python

import sys
import hashlib
import random
import crypt
import getpass

def salt(n):
	ch = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890./"
	salt = ""
	for i in range(0, n):
		s = random.randint(0,63)
		salt = salt + ch[s:s+1]
	return salt

print("Password Demo");
p = getpass.getpass("Enter your password: ");
print("Your password is [{}]".format(p));

if len(sys.argv) > 1:
	s_des = sys.argv[1][0:2]
	s_md5 = "$1$" + sys.argv[1][0:8] + "$";
	s_sha256 = "$5$" + sys.argv[1][0:8] + "$";
	s_sha512 = "$6$" + sys.argv[1][0:8] + "$";
else:
	s_des = salt(2);
	s_md5 = "$1$" + salt(8) + "$";
	s_sha256 = "$5$" + salt(8) + "$";
	s_sha512 = "$6$" + salt(8) + "$";

p_des = crypt.crypt(p, s_des)
p_md5 = crypt.crypt(p, s_md5)
p_sha256 = crypt.crypt(p, s_sha256)
p_sha512 = crypt.crypt(p, s_sha512)

print("    DES: {:<2d} {}".format(len(p_des) - 2, p_des));
print("    MD5: {:<2d} {}".format(len(p_md5) - 12, p_md5));
print(" SHA256: {:<2d} {}".format(len(p_sha256) - 12, p_sha256));
print(" SHA512: {:<2d} {}".format(len(p_sha512) - 12, p_sha512));

